include/envoy/buffer/buffer.h

```diff
@@ -249,7 +249,7 @@ class Instance {
     const auto sign_extension_bits =
         std::is_signed<T>::value && Size < sizeof(T) && bytes[most_significant_read_byte] < 0
             ? static_cast<T>(static_cast<typename std::make_unsigned<T>::type>(all_bits_enabled)
-                             << (Size * CHAR_BIT))
+                             << ((Size % sizeof(T)) * CHAR_BIT))
```

Comment:
The error is: `envoy/buffer/buffer.h(253): warning C4293: '<<': shift count negative or too big, undefined behavior` when compiling the template: `peekInt<T,ByteOrder::LittleEndian,4>(uint64_t)` with `T=int32_t `.

The problem here is that the left shift here will be 32 bits on a 32 bit integer, which is undefined. Now, we know that this branch of the conditional will not be taken (since `Size == sizeof(T)`), but MSVC is still compiling the branch. (If I understand `constexpr if` correctly, it would help here, but I think that's C++17).

The mod here is so that  when `Size < sizeof(T)`, we shift the correct amount, and when `Size == sizeof(T)` we don't generate an invalid left shift

Reply:
I see, can you add a comment? This is very interesting but not very intuitive. :)